pkg/binding/selectors.go

```diff
@@ -117,17 +74,104 @@ func (c *Controller) updateDecisions(obj runtime.Object) error {
-				bindingpolicy.GetName(), util.RefToRuntimeObj(obj), err)
+				bindingPolicy.GetName(), util.RefToRuntimeObj(obj), err)
 		}
 
 		if resolutionUpdated {
 			// enqueue binding to be synced since an object was added to its bindingpolicy's resolution
 			c.logger.V(4).Info("enqueued Binding for syncing due to a noting of an "+
-				"object in its resolution", "binding", bindingpolicy.GetName(),
+				"object in its resolution", "binding", bindingPolicy.GetName(),
 				"object", util.RefToRuntimeObj(obj))
-			c.enqueueBinding(bindingpolicy.GetName())
+			c.enqueueBinding(bindingPolicy.GetName())
+		}
+
+		// make sure object has singleton status if needed
+		if !isBeingDeleted(obj) &&
+			c.bindingPolicyResolver.ResolutionRequiresSingletonReportedState(bindingPolicy.GetName()) {
+			if err := c.addSingletonLabel(ctx, obj, bindingPolicy.GetName()); err != nil {
+				return fmt.Errorf("failed to add singleton label to object: %w", err)
+			}
+
+			isSelectedBySingletonBinding = true
+		}
+	}
+
+	// if the binding-policies matching cycles end and the object is not selected by a singleton binding,
+	// we need to remove the singleton label from the object if it exists.
+	// NOTE that this takes care of the case where the object was previously selected by a singleton binding
+	// and is no longer selected by any binding.
+	if !isSelectedBySingletonBinding {
+		if err := c.removeSingletonLabel(ctx, obj); err != nil {
+			return fmt.Errorf("failed to remove singleton label from object: %w", err)
 		}
 	}
 
 	return nil
 }
+
+func (c *Controller) removeSingletonLabel(ctx context.Context, obj runtime.Object) error {
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
+
+	labels := unstructuredObj.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	// skip if label does not exist
+	if _, found := labels[util.BindingPolicyLabelSingletonStatusKey]; !found {
+		return nil
+	}
+
+	// label found, should delete it and update object
+	delete(labels, util.BindingPolicyLabelSingletonStatusKey)
+	unstructuredObj.SetLabels(labels)
+
+	gvr, found := c.gvkGvrMapper.GetGvr(unstructuredObj.GetObjectKind().GroupVersionKind())
+	if !found {
+		return fmt.Errorf("failed to get GVR for object %v", util.RefToRuntimeObj(obj))
+	}
+
+	if unstructuredObj.GetNamespace() == metav1.NamespaceNone {
+		_, err := c.dynamicClient.Resource(gvr).Update(ctx, unstructuredObj, metav1.UpdateOptions{})
+		return err
+	}
+
+	_, err := c.dynamicClient.Resource(gvr).Namespace(unstructuredObj.GetNamespace()).Update(ctx, unstructuredObj, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *Controller) addSingletonLabel(ctx context.Context, obj runtime.Object, bindingPolicyName string) error {
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
```

Comment:
FYI: We know at development time that `ok` will always be `true`, right? If so then it would be legitimate to replace this with simply

```suggestion
	unstructuredObj := obj.(*unstructured.Unstructured)
```